### PR TITLE
[query] Add tag filters for query results from config

### PR DESCRIPTION
**Pull request by arnikola**
**What this PR does / why we need it**:
This adds filter parameters to query, based on a name and a list of regexes describing which values to remove from output.

**Special notes for your reviewer**:
Needs:
 - [x] tests
- [x] plumbing through options
- [x] parsing config

### CODE REVIEW

File: src/query/models/types.go [hunk 1]
```diff
@@ -150,3 +163,14 @@ type Metric struct {
...
+
+// Filters is a set of tag filters.
+type Filters []Filter
+
+// Filter is a regex tag filter.
+type Filter struct {
+	// Name is the name of the series.
+	Name []byte
+	// Filters are a set of regex expressions that filter matched tag values.
+	Filters []*regexp.Regexp
```
Comment (robskillington) [inline src/query/models/types.go:175]: Hm, perhaps we should also have some exact match ones here too (i.e. can specify either regexp or exact match)? There's cases where we just want exact match (in fact primarily now it's that) and ideally we wouldn't run regexp.

Comment (arnikola) [reply src/query/models/types.go:175]: Figured we'd handle this with anchored regex matchers?

Comment (robskillington) [reply src/query/models/types.go:175]: Yeah I'm just more concerned about if we only do exact matchers how much slower `bytes.Compare(.., ...) == 0` is vs `regexp.Match`?  Mainly for thousands of timeseries use case.

File: src/query/storage/m3/consolidators/filter.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+	return filteredTags
+}
+
+func filterTags(tags []CompletedTag, filters models.Filters) []CompletedTag {
+	if len(filters) == 0 || len(tags) == 0 {
+		return tags
+	}
+
+	skip := false
```
Comment (shreyassrivatsan) [inline src/query/storage/m3/consolidators/filter.go:98]: same as above

File: src/query/storage/m3/consolidators/filter.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+		name := tag.Name.Bytes()
+		value := tag.Value.Bytes()
+		for _, f := range filters {
+			if !bytes.Equal(name, f.Name) {
+				continue
+			}
+
+			// 0 length filters implies filtering for entire range.
+			if len(f.Values) == 0 {
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/filter.go:48]: Nice 👍

File: src/query/storage/m3/consolidators/filter.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+package consolidators
+
+import (
+	"bytes"
+
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/x/ident"
+)
+
+func filterTagIterator(iter ident.TagIterator, filters models.Filters) bool {
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/filter.go:30]: Should this perhaps return (bool, error) and check iter.Err() at the end?

File: src/query/storage/m3/consolidators/filter.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	return false
+}
+
+func filterNames(tags []CompletedTag, filters models.Filters) []CompletedTag {
+	if len(filters) == 0 || len(tags) == 0 {
+		return tags
+	}
+
+	filteredTags := tags[:0]
+	skip := false
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/query/storage/m3/consolidators/filter.go:69]: +1, skip not used outside of loop

File: src/query/storage/m3/consolidators/filter.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	for _, tag := range tags {
+		skip = false
+		for _, f := range filters {
+			if len(f.Values) != 0 {
+				// If this has filter values, it is not a name filter, and the result
+				// is valid.
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/filter.go:75]: Hm wouldn't we want to skip this tag name though if tag.Name matches f.Name?  i.e. if we don't want to see any time series with "foo_k8s_namespace" then we don't want the "foo_k8s_namespace" tag name to come back either right?

File: src/query/storage/m3/consolidators/filter.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (arnikola) [reply src/query/storage/m3/consolidators/filter.go]: 👍

File: src/query/storage/m3/consolidators/filter.go (hunk 4 above)
Comment (arnikola) [reply src/query/storage/m3/consolidators/filter.go:30]: Ah yeah, good call 👍

Comment (arnikola) [reply src/query/storage/m3/consolidators/filter.go:30]: Hm makes it a bit weird with a defer, will refactor a little

**Pull request merged**